### PR TITLE
Login/Register Component

### PR DESCRIPTION
**Pull request by TheRealJessicaLi**
## Description
Tickets:
[Signup UI](https://www.notion.so/uwblueprintexecs/Sign-up-Frontend-UI-285cd83e4ac04457882e80cd9b44cf26)
[Login UI](https://www.notion.so/uwblueprintexecs/Login-Frontend-UI-fc6e23155665482db710f5091a588440)
[Combining into one](https://www.notion.so/uwblueprintexecs/Combine-login-and-sign-up-on-one-page-7e7d9406f94b4afaaf56b1928dbe8667)

Built the login/signup component so they are both on the same route. Forgot password link is not yet set up and error states are not implemented in these tickets, but login/register should work.

## Testing

- Pull this branch and run the app locally
- Navigate to [http://localhost:3000/login](http://localhost:3000/login) and make sure the UI matches/behaviour is correct

### Tablet
Screen Shot 2021-02-28 at 2 36 06 PM

### Mobile
Screen Shot 2021-02-28 at 2 36 37 PM

### CODE REVIEW

File: src/shared/App.js [hunk 1]
```diff
@@ -45,10 +43,7 @@ function App() {
           <Switch>
             {/* TODO: The login and register should be combined to a single path as per the design */}
```
Comment (jzxhuang) [inline src/shared/App.js]: can remove this comment

File: src/components/LoginRegister/LoginInput.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+  const onChangePass = e => {
+    setPassword(e.target.value);
+  };
+
+  return (
+    <Box>
+      <VStack spacing="24px">
+        <Flex w="100%" justify="left">
+          <Heading as="h1">Welcome lorem</Heading>
```
Comment (jzxhuang) [inline src/components/LoginRegister/LoginInput.js]: I think this text should be just "Welcome" 😂

File: src/components/LoginRegister/LoginInput.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        </Flex>
+        <Input type="email" name="email" placeholder="Email" value={email} onChange={onChangeEmail} />
+        <Input
+          type="password"
+          name="password"
+          placeholder="Password"
+          value={password}
+          onChange={onChangePass}
+          mb="8px"
```
Comment (jzxhuang) [inline src/components/LoginRegister/LoginInput.js]: would be great if we could add handler for pressing "Enter" to login, easiest way might be to wrap in a `form` component like in #35 (not sure what form components Chakra provides)

File: src/components/LoginRegister/LoginInput.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+const LoginInput = () => {
+  const client = useApolloClient();
+  const [email, setEmail] = useState('');
+  const [password, setPassword] = useState('');
+  const dispatch = useDispatch();
+
+  const handleClick = async e => {
+    e.preventDefault();
+    dispatch(login(email, password, client));
```
Comment (jzxhuang) [inline src/components/LoginRegister/LoginInput.js]: not sure how much work this is but would be great if we can add in simple error message if the login fails (if it's too much refactoring we can do this in another PR)

i.e. if the login response is `null` show a generic error message like "Oops! Check your credentials and try again."

File: src/components/LoginRegister/RegisterInput.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  const client = useApolloClient();
+  const [email, setEmail] = useState('');
+  const [password, setPassword] = useState('');
+  const [firstName, setFirstName] = useState('');
+  const [lastName, setLastName] = useState('');
+  const dispatch = useDispatch();
+
+  const handleClick = async e => {
+    e.preventDefault();
+    dispatch(register(firstName, lastName, email, password, client));
```
Comment (jzxhuang) [inline src/components/LoginRegister/RegisterInput.js]: Not sure how much work this is but after registering successfully we should redirect to the home page? if too much work do in another PR 

also a simple error message handler like the comment for login component

File: src/pages/User/LoginRegister.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+                  outline: 'none',
+                }}
+                _hover={{
+                  color: 'brand.50',
+                }}
+                onClick={() => setLoginRegister('login')}
+              >
+                Sign In
+              </Button>
+              {loginRegister === 'login' && <Box bg="brand.50" h="5px" borderRadius="sm"></Box>}
```
Comment (jzxhuang) [inline src/pages/User/LoginRegister.js]: this is a nit but best code quality would probably be to abstract both the Button and the Box into components (i.e. `TabButton` and `TabUnderline`, probably components in the same file is ok

File: src/pages/User/LoginRegister.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
+import React, { useState } from 'react';
+import { Box, Flex, Button, Image } from '@chakra-ui/react';
+import LoginInput from '../../components/LoginRegister/LoginInput';
+import RegisterInput from '../../components/LoginRegister/RegisterInput';
+import logo from '../../images/logo-black.png';
+
+function LoginRegister() {
+  const [loginRegister, setLoginRegister] = useState('login');
```
Comment (jzxhuang) [inline src/pages/User/LoginRegister.js]: Prefer caps for constant state strings, i.e. `LOGIN` and `REGISTER`

**Review by jzxhuang — COMMENTED** [review]
ur responsive design is pretty sick

**Comment by jzxhuang** [comment]
Woahh ok cynthia also found this super lit `<FormControl>` component from Chakra that seems really useful, see #35

File: src/components/LoginRegister/LoginInput.js (hunk 4 above)
Comment (TheRealJessicaLi) [reply src/components/LoginRegister/LoginInput.js]: Added a simple error message for both login and register!

File: src/components/LoginRegister/RegisterInput.js (hunk 5 above)
Comment (TheRealJessicaLi) [reply src/components/LoginRegister/RegisterInput.js]: Done!

**Comment by TheRealJessicaLi** [comment]
> Woahh ok cynthia also found this super lit `<FormControl>` component from Chakra that seems really useful, see #35

That looks really cool! That might be useful for input validation, but I think that should be a separate ticket since there's a lot of error states to cover

File: src/components/LoginRegister/LoginInput.js (hunk 3 above)
Comment (TheRealJessicaLi) [reply src/components/LoginRegister/LoginInput.js]: Yep! I ended up wrapping everything in a `form`

File: src/components/LoginRegister/RegisterInput.js [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+  return (
+    <form onSubmit={e => handleClick(e)}>
+      <VStack spacing="24px">
+        <Flex w="100%" justify="left">
+          <Heading as="h1">Create an account</Heading>
+        </Flex>
+        <Input name="firstName" placeholder="First Name" value={firstName} onChange={onChangeFirstName} />
+        <Input name="lastName" placeholder="Last Name" value={lastName} onChange={onChangeLastName} />
+        <Input name="email" placeholder="Email" value={email} onChange={onChangeEmail} />
```
Comment (jzxhuang) [inline src/components/LoginRegister/RegisterInput.js]: add `type="email"` here and you actually get some default form validation! It doesn't validate "not empty" but it does validate you enter a valid email format if you do start typing something :O 

image

```suggestion
        <Input type="email" name="email" placeholder="Email" value={email} onChange={onChangeEmail} />
```

File: src/components/LoginRegister/LoginInput.js [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+  const dispatch = useDispatch();
+
+  const handleClick = async e => {
+    e.preventDefault();
+    dispatch(login(email, password, client)).then(
+      loggedIn => {
+        if (!loggedIn) setSubmitState('FAIL');
+      },
+      error => {
+        setSubmitState('FAIL');
```
Comment (jzxhuang) [inline src/components/LoginRegister/LoginInput.js:21]: just curioussss do you know if this is some "ESNext" shorthand for `.then` and `.catch`? I see this pattern throughout our codebase but I don't think I've ever seen it before anywhere else / I can't find the docs for this

Just curious if you know or you just copied the pattern 🤷

File: src/components/LoginRegister/RegisterInput.js [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+  const handleClick = async e => {
+    e.preventDefault();
+    dispatch(register(firstName, lastName, email, password, client)).then(
+      registered => {
+        if (!registered) {
+          setSubmitState('FAIL');
+        } else {
+          // Redirect to home page after registering
+          return <Redirect to={{ pathname: '/home' }} />;
+        }
```
Comment (jzxhuang) [inline src/components/LoginRegister/RegisterInput.js]: we shouldn't actually need this `Redirect` due to you updating `REGISTER_SUCCESS` correctly in the Redux reducer and L45 in `App.js` (Redirects away from login page to home page if the `user` object is not null)

```suggestion
        if (!registered) {
          setSubmitState('FAIL');
        }
```

**Review by jzxhuang — COMMENTED** [review]
noice let's get a design review! left very minor comments 

could you follow up with shaahana/stacy on https://uwblueprint.slack.com/archives/C01JT2AE91R/p1614571636006000?thread_ts=1614563948.003700&cid=C01JT2AE91R and ask them to take a look?

Send the Netlify deploy preview link and what to test (i.e. the whole login/register flow), as well as any notable differences from the Figma design

It may also be worth mentioning that there are some logical errors not related to the UI/UX (i.e. you can create an account with completely empty name/email/password), I think I accidentally created an account in prod there the email/pw are both `""` 😂 

Also may be worth mentioning to them that to "logout" you need to go to `/logout` on this preview branch

**Comment by jzxhuang** [comment]
loool sorry for random follow up comments but we can add `isRequired` to every form field to quickly add in some simple front end form validation (to cover up the broken backend)

see https://chakra-ui.com/docs/form/input search for `isRequired` I say we just add that to every single input field in Login and Register :D

File: src/components/LoginRegister/LoginInput.js (hunk 9 above)
Comment (TheRealJessicaLi) [reply src/components/LoginRegister/LoginInput.js:21]: Yeah, it's syntax for handling a Promise (I read about it [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise#chained_promises)). The first argument is a callback for the resolved case, and the second is for the rejected case. I think the alternative is to chain `.then` and `.catch`, but they should do the same thing!

File: src/components/LoginRegister/RegisterInput.js (hunk 8 above)
Comment (TheRealJessicaLi) [reply src/components/LoginRegister/RegisterInput.js]: Done!

File: src/components/LoginRegister/RegisterInput.js (hunk 10 above)
Comment (TheRealJessicaLi) [reply src/components/LoginRegister/RegisterInput.js]: Removed!

**Comment by TheRealJessicaLi** [comment]
> loool sorry for random follow up comments but we can add `isRequired` to every form field to quickly add in some simple front end form validation (to cover up the broken backend)
> 
> see https://chakra-ui.com/docs/form/input search for `isRequired` I say we just add that to every single input field in Login and Register :D

All good, updated!

**Comment by TheRealJessicaLi** [comment]
> noice let's get a design review! left very minor comments
> 
> could you follow up with shaahana/stacy on https://uwblueprint.slack.com/archives/C01JT2AE91R/p1614571636006000?thread_ts=1614563948.003700&cid=C01JT2AE91R and ask them to take a look?
> 
> Send the Netlify deploy preview link and what to test (i.e. the whole login/register flow), as well as any notable differences from the Figma design
> 
> It may also be worth mentioning that there are some logical errors not related to the UI/UX (i.e. you can create an account with completely empty name/email/password), I think I accidentally created an account in prod there the email/pw are both `""` 😂
> 
> Also may be worth mentioning to them that to "logout" you need to go to `/logout` on this preview branch

Pinged them for a design review!

File: src/components/LoginRegister/LoginInput.js (hunk 9 above)
Comment (jzxhuang) [reply src/components/LoginRegister/LoginInput.js:21]: that is pre cool i feel like i am so behind on JS syntax there is new syntax every year lolol

File: src/components/LoginRegister/LoginInput.js [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+      {submitState === 'FAIL' && (
+        <Flex w="100%" justify="left">
+          <p>Oops! Check your credentials and try again.</p>
+        </Flex>
+      )}
+      <Flex justify="center" mt="24px">
+        <Button role="link" width="131px" height="43px" type="submit">
+          Sign In
+        </Button>
+      </Flex>
```
Comment (jzxhuang) [inline src/components/LoginRegister/LoginInput.js]: ok so I just played around w/ Chakra and styling a bit more 

* Move `VStack` to wrap everything (just like in `RegisterInput`)
* Use `alignSelf` rather than creating a new `Flex` component (because `VStack itself creates a flex with align-items: center`
* For boolean props, you can just pass the prop name to set it to true, so `isRequired` rather than `isRequired={true}`

```suggestion
      <VStack spacing="24px">
        <Heading alignSelf="flex-start" size="h1" as="h1">
          Welcome
        </Heading>
        <Input type="email" name="email" placeholder="Email" value={email} onChange={onChangeEmail} isRequired />
        <Input
          type="password"
          name="password"
          placeholder="Password"
          value={password}
          onChange={onChangePass}
          mb="8px"
          isRequired
        />
        <Link alignSelf="flex-start">Forgot Password?</Link>
        {submitState === 'FAIL' && <Text alignSelf="flex-start">Oops! Check your credentials and try again.</Text>}
        <Button role="link" width="131px" height="43px" type="submit">
          Sign In
        </Button>
      </VStack>
```

end result is it still looks the same lool

image

File: src/components/LoginRegister/RegisterInput.js [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+  };
+
+  const onChangePass = e => {
+    setPassword(e.target.value);
+  };
+
+  return (
+    <form onSubmit={e => handleClick(e)}>
+      <VStack spacing="24px">
+        <Flex w="100%" justify="left">
```
Comment (jzxhuang) [inline src/components/LoginRegister/RegisterInput.js]: could you make the same changes here as the `LoginInput` (so remove flexbox and use `alignSelf`

File: src/components/LoginRegister/RegisterInput.js [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+          <Heading as="h1">Create an account</Heading>
+        </Flex>
+        <Input
+          name="firstName"
+          placeholder="First Name"
+          value={firstName}
+          onChange={onChangeFirstName}
+          isRequired={true}
```
Comment (jzxhuang) [inline src/components/LoginRegister/RegisterInput.js]: same changes for `isRequired`

File: src/components/LoginRegister/RegisterInput.js [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+          value={password}
+          onChange={onChangePass}
+          isRequired={true}
+        />
+        {submitState === 'FAIL' && (
+          <Flex w="100%" justify="left">
+            <p as="subtitle" size="subtitle">
+              Unable to register, please try again.
+            </p>
+          </Flex>
```
Comment (jzxhuang) [inline src/components/LoginRegister/RegisterInput.js]: I think here we can just use a Chakra `Text` with default size and `alignSelf="flex-start"` like the suggested change in `LoginInput`

**Review by jzxhuang — COMMENTED** [review]
i swr i don't actually care about CSS semantics that much but I noticed u structured some of the code in `LoginInput` differently from `RegisterInput` so I wanted to fix that and fixed some styling along the way

There's probably better semantics you could do in `LoginRegister.js` but like I'm too lazy to look for that so let's make these fixes then merge (after designers take a look too)

File: src/data/reducers/auth.js [hunk 15]
```diff
@@ -14,21 +14,29 @@ const initialState = {
 
 export default function (state = initialState, action) {
   const { type, payload } = action;
+  let firstName, lastName, email, userId, teamId;
```
Comment (patrick-du) [inline src/data/reducers/auth.js]: Nit: would it make sense to define these as constants inside of each case? 

Ex. `const { firstName, lastName, email, userId, teamId } = payload;`

File: src/data/reducers/auth.js [hunk 16]
```diff
@@ -14,21 +14,29 @@ const initialState = {
...
         ...state,
         // TODO: email, userId, firstName, lastName, should be set here
-        user: null,
+        user: {
+          payload,
+          lastName,
+          email,
+          userId,
+          teamId,
+        },
```
Comment (patrick-du) [inline src/data/reducers/auth.js]: I think this stores a payload object within the user object and then sets the same fields in payload into the user object again which ends up looking like this:
```
user: {
    payload: {
        firstName: 'asd',
	lastName: 'fgh',
	email: 'asd@gmail.com',
	userId: 123,
	teamId: 456
    },
    lastName: 'fgh',
    email: 'asd@gmail.com',
    userId: 123,
    teamId: 456
}
```

Not sure if I'm understanding this correctly or if the above is what we want but if payload is everything we expect to be stored in the user object, we can pass it directly which ends up looking like this:

```
user: {
    firstName: 'asd',
    lastName: 'fgh',
    email: 'asd@gmail.com',
    userId: 123,
    teamId: 456
}
```

```suggestion
        user: payload
```

File: src/data/reducers/auth.js [hunk 17]
```diff
@@ -14,21 +14,29 @@ const initialState = {
...
 export default function (state = initialState, action) {
   const { type, payload } = action;
+  let firstName, lastName, email, userId, teamId;
 
   switch (type) {
     case REGISTER_SUCCESS:
+      ({ firstName, lastName, email, userId, teamId } = payload);
       return {
         ...state,
         // TODO: email, userId, firstName, lastName, should be set here
```
Comment (patrick-du) [inline src/data/reducers/auth.js]:
```suggestion
```

**Comment by patrick-du** [comment]
Looks really good 😄 - just left minor comments to clarify/confirm my own understanding of some changes

File: src/components/LoginRegister/LoginInput.js (hunk 11 above)
Comment (TheRealJessicaLi) [reply src/components/LoginRegister/LoginInput.js]: Fixed, thanks!

File: src/components/LoginRegister/RegisterInput.js (hunk 12 above)
Comment (TheRealJessicaLi) [reply src/components/LoginRegister/RegisterInput.js]: Done!

File: src/components/LoginRegister/RegisterInput.js (hunk 13 above)
Comment (TheRealJessicaLi) [reply src/components/LoginRegister/RegisterInput.js]: Done!

File: src/components/LoginRegister/RegisterInput.js (hunk 14 above)
Comment (TheRealJessicaLi) [reply src/components/LoginRegister/RegisterInput.js]: Done!

File: src/data/reducers/auth.js (hunk 15 above)
Comment (TheRealJessicaLi) [reply src/data/reducers/auth.js]: Yeah I agree and I tried that, but I was getting an error that those constants were already defined in the first switch case, so I ended up going with the let statement (not sure if there's a better way). I just took this out and went with your approach in the next comment though, so this won't be a problem. Thanks!

File: src/data/reducers/auth.js (hunk 16 above)
Comment (TheRealJessicaLi) [reply src/data/reducers/auth.js]: Good point, thanks!

File: src/data/reducers/auth.js (hunk 17 above)
Comment (TheRealJessicaLi) [reply src/data/reducers/auth.js]: Fixed!

**Review by patrick-du — APPROVED** [review]
LGTM! 👍

**Review by jzxhuang — APPROVED** [review]
🚀
